Created a chapter on reproduciblity in AI. (#4681)
the-turing-way/the-turing-way: book/website/reproducible-research/reproducibility-in-ai.md

For researchers, this raises questions about how AI applies to science and how we can create AI systems that support the principle of reproducibility. Reproducibility is particularly important in AI because AI research often involves many interacting components, including data, models, code, computational environments, and stochastic processes.

Reproducible research helps researchers examine, verify, and build on existing work. It provides a record of how results were produced, making it easier to identify errors, understand the basis for results, and avoid unnecessary duplication of effort.

In AI research, these principles are particularly important because the final result can depend on many different parts of the research process. Changes to the training data, code, model configuration, software, hardware, or computational environment can affect the resulting model or its outputs.

Understanding how these components affect reproducibility can therefore help researchers build on existing AI research and make its results easier to study and verify.

> We write about the benefits of reproducibility more extensively in the [Added Advantages](https://book.the-turing-way.org/reproducible-research/overview/overview-benefit/) chapter of *The Turing Way*.

Reviewer: We should use a cross-reference here.

> ### What is an AI system?

Reviewer: I still don't think this is the place to define AI

Before discussing reproducibility in AI, it is useful to establish what we mean by an AI system. There are several definitions of artificial intelligence, two of which are highlighted here.

> **Definition 1**

Reviewer: We shouldn't use bold text like this. It is unfriendly for those using screen readers and makes navigation more difficult.

>
> It is a field of research in engineering, mathematics and computer science that develops and studies methods and software that enable machines to perceive their environment and use learning and intelligence to take actions that maximize their chances of achieving defined goals.
>
> -- [Wikipedia](https://en.wikipedia.org/wiki/Artificial_intelligence#)

**Definition 2**

> An AI system is a machine-based system that, for explicit or implicit objectives, infers, from the input it receives, how to generate outputs such as predictions, content, recommendations, or decisions that can influence physical or virtual environments.
>
> Different AI systems vary in their levels of autonomy and adaptiveness after deployment, as defined by the OECD.
>
> -- [Organisation for Economic Co-operation and Development (OECD)](https://www.oecd.org/content/dam/oecd/en/publications/reports/2024/03/explanatory-memorandum-on-the-updated-oecd-definition-of-an-ai-system_3c815e51/623da898-en.pdf)

For the purposes of this chapter, we use the OECD definition of an AI system [@oecd2024ai].

> ### What does "open source AI" mean?

Reviewer: Similar to above, I don't think this is the right place to define open source AI.

When discussing reproducibility, it is useful to consider how it relates to openness. Although reproducibility and openness are not the same thing, they can overlap.

With AI, there is no universally agreed definition of what makes an AI system open source. Different definitions place emphasis on different components of an AI system. The discussion is usually about which structural elements of an AI system should be made available, how much of them should be available, and whether a single definition can adequately describe openness across different types of AI systems.

Several definitions and frameworks have been proposed.

For example, the [OSI Open Source AI Definition](https://opensource.org/ai/open-source-ai-definition) frames openness around the freedoms to use, study, modify, and share an AI system.

Other approaches place greater emphasis on the availability of training information and training code, and some also consider the availability of training data. One example is the [Open Weights definition](https://github.com/Open-Weights/Definition/blob/main/Definition.md).

> The [_Opening up ChatGPT_](https://opening-up-chatgpt.github.io/) resource, maintained by Liesenfeld, Lopez, and Dingemanse, provides an updated overview of models considered open source (Liesenfeld et al., 2023).

Reviewer: ```suggestion
The [_Opening up ChatGPT_](https://opening-up-chatgpt.github.io/) resource, maintained by Liesenfeld, Lopez, and Dingemanse, provides an updated overview of models considered open source [@liesenfeldopening].
```

For this chapter, we do not adopt one definition of open-source AI over another. Instead, we consider the different components of an AI system that may be relevant to understanding and reproducing it.

These components can include:

- model parameters, including weights and biases;
- model architecture;
- training and inference code;
- training methodology and configuration;
- intermediate checkpoints;
- training datasets;
- training-data composition, provenance, and preprocessing; and
- software, hardware, and computational environments.

These components can be available to different degrees. For example, a model may make its weights available while keeping its training data or training code unavailable. Having access to one component therefore does not necessarily mean that the system can be fully studied or reproduced.

> ### What is reproducibility?

Reviewer: ```suggestion
### What is reproducibility with regards to AI?
```

> For this chapter, we use the definition of reproducibility described by *The Turing Way*: research that can be independently recreated using the same data and code as the original work.

Reviewer: It would be better to reuse/embed this rather than copy the text here.

We also draw on the work of Gundersen, Coakley, et al. [@gundersen2022sources], which discusses different sources and forms of reproducibility in machine learning.

More specifically:

> > Interpretation reproducibility does not require the reproduced experiment to have the same or similar outcome nor analysis, but requires the interpretation to be the same as the original one.
> >
> > -- Gundersen, Coakley, et al. (2022)

Reviewer: ```suggestion
> Interpretation reproducibility does not require the reproduced experiment to have the same or similar outcome nor analysis, but requires the interpretation to be the same as the original one.
>
> -- @gundersen2022sources
```

1. Gundersen, O. E., Coakley, K., & others. (2022). *Sources of irreproducibility in machine learning*. [Add the DOI or URL used in your existing `@gundersen2022sources` reference.]

1. Liesenfeld, A., Lopez, A., & Dingemanse, M. (2023). Opening up ChatGPT: Tracking openness, transparency, and accountability in instruction-tuned text generators. In *CUI '23: Proceedings of the 5th International Conference on Conversational User Interfaces*. 10.1145/3571884.3604316

1. Open Source Initiative. (2024). *The Open Source AI Definition 1.0*. [https://opensource.org/ai/open-source-ai-definition](https://opensource.org/ai/open-source-ai-definition)

1. Organisation for Economic Co-operation and Development (OECD). (2024). *Explanatory memorandum on the updated OECD definition of an AI system*. [https://www.oecd.org/content/dam/oecd/en/publications/reports/2024/03/explanatory-memorandum-on-the-updated-oecd-definition-of-an-ai-system_3c815e51/623da898-en.pdf](https://www.oecd.org/content/dam/oecd/en/publications/reports/2024/03/explanatory-memorandum-on-the-updated-oecd-definition-of-an-ai-system_3c815e51/623da898-en.pdf)

> 1. The Turing Way Community. (2026). *The Turing Way: A handbook for reproducible, ethical and collaborative research*. [https://book.the-turing-way.org/](https://book.the-turing-way.org/)

Reviewer: If these are references above we don't need to write them out. A references section will be created by MyST.

> # Reproducibility in AI

Reviewer: ```suggestion
---
abbreviations:
  OSI: Open Source Initiative
---
# Reproducibility in AI
```

At the time of this writing, AI has become a large part of human life and part of our daily processes.
Active and inactive users alike have to interact with AI, either directly or indirectly.
For researchers, the question arises of how this applies to science and how we can create AI systems that carry on the principle of reproducibility.
The topic of reproducibility in AI is important, as it helps us build on systems and research that can be reproduced and studied by researchers.

## Definitions

> ### What are AI systems

Reviewer: There might be a better place in the book for this. Not sure if it already exists though.

Author: I think, depending on how many chapters I am able to write, we might have a "Guide to Reproducibility in AI" page like the other guides. It might fit under there.

Reviewer 2: Hi @Preshh0 - you mean a landing page for the chapter? I think a guide to reproducibility in AI is a bit of an overkill - it fits perfectly within the reproducibility guide.

Author: I meant if there are other chapters in the future within the same theme of reproducibility in AI that might require it, so that they can be found in one place - just in case. 

That aside, there's no need to create a different page right now. You're right that it fits under the reproducibility guide.

Reviewer: It might not of been clear what I was saying. I mean the more general information about AI like "What is an AI system" should be somewhere else and not on the "reproducible AI" page.

Otherwise, I think this information is very likely (or at least more likely) to end up being duplicated.

It also might be more appropriate to define any terms we need for this page in the glossary.

Author: Ah, I get you. That's a good chapter I can work on. 

Since in this case, it was added to provide context for this chapter's topic, the definition can be linked here once it is written.

I added some terms to the glossary - not sure how much is too much, but I defined some terms specific to AI.

There are two definitions to be highlighted in this article.

**Wikipedia**
> Artificial intelligence (AI) is the capability of computational systems to perform tasks typically associated with human intelligence, such as learning, reasoning, problem-solving, perception, and decision-making.

> > It is a field of research in engineering, mathematics and computer science that develops and studies methods and software that enable machines to perceive their environment and use learning and intelligence to take actions that maximize their chances of achieving defined goals ([Wikipedia](https://en.wikipedia.org/wiki/Artificial_intelligence#)).

Reviewer: ```suggestion
> Artificial intelligence (AI) is the capability of computational systems to perform tasks typically associated with human intelligence, such as learning, reasoning, problem-solving, perception, and decision-making.
> It is a field of research in engineering, mathematics and computer science that develops and studies methods and software that enable machines to perceive their environment and use learning and intelligence to take actions that maximize their chances of achieving defined goals 
>
> -- [Wikipedia](https://en.wikipedia.org/wiki/Artificial_intelligence#).
```

Quotations in MyST support [specifying an author](https://mystmd.org/guide/typography#quotations). That should look better and be more accessible than using bold text.

**Organisation for Economic Co-operation and Development (OECD)**
> An AI system is a machine-based system that, for explicit or implicit objectives, infers, from the input it receives, how to generate outputs such as predictions, content, recommendations, or decisions that can influence physical or virtual environments.

> > Different AI systems vary in their levels of autonomy and adaptiveness after deployment. (OECD, Artificial Intelligence Papers, March 2024)

Reviewer: ```suggestion
> An AI system is a machine-based system that, for explicit or implicit objectives, infers, from the input it receives, how to generate outputs such as predictions, content, recommendations, or decisions that can influence physical or virtual environments.
> Different AI systems vary in their levels of autonomy and adaptiveness after deployment. 
>
> -- OECD, Artificial Intelligence Papers, March 2024
```

Can we add a link for this?

For the sake of this chapter, we lean more towards the OECD definition.

When discussing reproducibility, although it is not always the same definition as "open," there is a tendency to have open principles, as they sometimes overlap.

> In that regard, we define AI with respect to Open Source, using the Open Source Initiative [(OSI) definition](https://opensource.org/ai/open-source-ai-definition).

Reviewer: ```suggestion
In that regard, we define AI with respect to Open Source, using the OSI definition.
```

> In that regard, we define AI with respect to Open Source, using the Open Source Initiative [(OSI) definition](https://opensource.org/ai/open-source-ai-definition).

Reviewer: This feels tricky. _I_ think there are some problems with the OSI open source AI definition and I would imagine a lot of the community would feel the same way.

Maybe we should draw inspiration from it but not adopt it? And add some text about how in terms of reproducibility, training data is essential, which is [not strictly required by the OSI definition](https://opensource.org/ai/faq#why-do-you-allow-the-exclusion-of-some-training-data). And there is even the [open weights](https://opensource.org/ai/open-weights) concession where source code is not shared.

@EstherPlomp and @Arielle-Bennett may have opinions.

Reviewer 2: I also disagree with the OSI definition, so I'd prefer not to say that 'we agree' and instead provide it as an example as @JimMadge suggests

Author: I'd reword this, thank you.

@EstherPlomp, can I know what definition you prefer? Asking for extra references so I can provide a more informed text for that section.

Thank you.

Reviewer 2: My definition would be to have the data and code openly available - like @JimMadge already mentioned - the training data is important and if that is not open I don't think you can call it open AI.

Reviewer: I think it probably isn't for us to _make_ a definition. But if we are discussing what Open AI _is_ I don't think we can bring in the OSI definition without talking about the wider discussion around it.

Author: Reworked the whole chapter. Happy to see reviews whenever anyone has bandwidth.

Thank you!

> In this case, a 'system' encompasses both a fully functional structure and its discrete structural elements.
> To be considered Open Source, the requirements are the same, whether applied to a system, a model, weights and parameters, or other structural elements.
>
> An Open Source AI is an AI system made available under terms and in a way that grant the freedoms to:
> - Use the system for any purpose and without having to ask for permission.
> - Study how the system works and inspect its components.
> - Modify the system for any purpose, including to change its output.

> > - Share the system for others to use with or without modifications, for any purpose.

Reviewer: ```suggestion
> In this case, a 'system' encompasses both a fully functional structure and its discrete structural elements.
> To be considered Open Source, the requirements are the same, whether applied to a system, a model, weights and parameters, or other structural elements.
>
> An Open Source AI is an AI system made available under terms and in a way that grant the freedoms to:
> - Use the system for any purpose and without having to ask for permission.
> - Study how the system works and inspect its components.
> - Modify the system for any purpose, including to change its output.
> - Share the system for others to use with or without modifications, for any purpose.
>
> -- [OSI open source AI definition](https://opensource.org/ai/open-source-ai-definition)
```

Finally, we define what reproducibility is according to The Turing Way, and narrow it down with the definition by Gundersen, Coakley, et al. (2023).

According to The Turing Way, reproducible research is work that can be independently recreated from the same data and the same code that the original team used.
More specifically:

> Interpretation reproducibility does not require the reproduced experiment to have the same or similar outcome nor analysis, but requires the interpretation to be the same as the original one.
>

> > — Gundersen, Coakley, et al. (2023)

Reviewer: ```suggestion
> Interpretation reproducibility does not require the reproduced experiment to have the same or similar outcome nor analysis, but requires the interpretation to be the same as the original one.
>
> -- Gundersen, Coakley, et al. (2023)
```

We should add this reference to the bibliography.

## Why is reproducibility important

Reproducibility is important because it ensures that research findings or experimental results can be repeated by others using the same methods, data, and procedures.
This helps verify the accuracy and reliability of the results, increasing confidence in the conclusions.
Reproducibility also promotes transparency by encouraging researchers to clearly document their methods, making it easier for others to understand, validate, and build upon existing work.
As a result, it supports scientific progress, reduces errors, and enables informed decision-making in fields such as science, medicine, engineering, and technology.

> In summary: open data and code, ease of access, and review.

Reviewer: We can probably defer to an existing chapter for this rather than writing similar text here.

Author: Done!

- **Set a random seed**: let your computer know to begin evaluations from the same point every time, to enable some level of predictability, so results are as near as possible to the original result.
- **Log all results**: the best way to avoid duplicating efforts is to know what didn't work and why it didn't work.
  All results should be logged, both positive and negative.
- **Work with a checklist**: when working, it can be easy to forget some of the important steps.
  Having a simple checklist enables you to know what's missing and what's not, without a lot of mental hassle.

### A simple comparison of models and systems

| Model / system | Weights | Training code | Training data | What you can do |
|---|---|---|---|---|
| Llama (Meta) | ✅ | ❌ | ❌ | Download and run, fine-tune |
| Mistral models | ✅ | ❌ | ❌ | Download and run, fine-tune |
| Gemma (Google) | ✅ | ❌ | ❌ | Download and run, fine-tune |
| BLOOM (BigScience) | ✅ | ✅ | ✅ | Fully reproduce training |
| OLMo (AI2) | ✅ | ✅ | ✅ | Fully reproduce training |
| GPT / Claude / Gemini (hosted API) | ❌ | ❌ | ❌ | Query via API only |

> *Table generated by Claude, Sonnet 5 (effort: medium).*

Reviewer: ```suggestion
:::{table} Table generated by Claude, Sonnet 5 (effort: medium).
:label: tab-ai-reproducibility
| Model / system | Weights | Training code | Training data | What you can do |
|---|---|---|---|---|
| Llama (Meta) | ✅ | ❌ | ❌ | Download and run, fine-tune |
| Mistral models | ✅ | ❌ | ❌ | Download and run, fine-tune |
| Gemma (Google) | ✅ | ❌ | ❌ | Download and run, fine-tune |
| BLOOM (BigScience) | ✅ | ✅ | ✅ | Fully reproduce training |
| OLMo (AI2) | ✅ | ✅ | ✅ | Fully reproduce training |
| GPT / Claude / Gemini (hosted API) | ❌ | ❌ | ❌ | Query via API only |
```

Have you checked that this is correct?

Are is the training really fully reproducible? Training is a stochastic process and AFAIK the results will depend on the hardware you use.

Reviewer 2: I think it would be better to refer to a valided resource than something generated by Claude about this, and this is already available here! https://opening-up-chatgpt.github.io/

Author: Thank you so much for this resource! I have enjoyed looking at, and learning from it.

Do you think it makes sense to leave the table in the chapter, or remove it entirely and just link the website so readers can reference it on their own. Especially if those models get updated as time goes on?

Author: I am of the opinion to remove it entirely, but very open to suggestions!

Reviewer 2: I think the external resource would be better as this is maintained by the people that started it so that would save us updating this page in the long run!

Author: Edited, thank you!

Most "open" models only tick the first box.
Full reproducibility needs all three.

### References

- Gundersen, O. E., Cappelen, O., Mølnå, M., & Nilsen, N. G. (2024). The unreasonable effectiveness of open science in AI: A replication study (arXiv No. 2412.17859). arXiv. https://doi.org/10.48550/arXiv.2412.17859 
- Pineau, J., Vincent-Lamarre, P., Sinha, K., Larivière, V., Beygelzimer, A., d'Alché-Buc, F., Fox, E., & Larochelle, H. (2020). Improving reproducibility in machine learning research: A report from the NeurIPS 2019 reproducibility program (arXiv No. 2003.12206). arXiv. https://doi.org/10.48550/arXiv.2003.12206

> - Gundersen, O. E., Coakley, K., Kirkpatrick, C. R., & Gil, Y. (2022). Sources of irreproducibility in machine learning: A review. arXiv. https://doi.org/10.48550/arXiv.2204.07610

Reviewer: ```suggestion
```

These should go in the bibliography and cited in the page https://mystmd.org/guide/citations#markdown-citations

Author: Done. Do I remove the references now from the chapter, or I can leave it still?
